### DISCUSSION TITLE
low DL throughput, but good UL throughput (AGW v1.6)

### DISCUSSION THREAD

**Discussion by alksara** [magma/magma#8953] [General]
We are observing low DL throughput (~3 Mbps), but good UL throughput (~18 Mbps) within our PoC. After looking at pcap, we see all packets received on sgi (with different lengths) transmitted with size 1416 when GTP encapsulated on s1u. our UE supports mtu 1400, however, we see gtp_br0 configured with mtu 1500 on the AGW. Is there a way to reduce mtu for gtp encapsulation to 1400 on AGW? pcap on sgi, s1-u, and LTE usb dongle attached. Also, our environment description is below. Any help is appreciated. Thanks.

Environment
**Version: 1.6
**Affected Component: AGW
**Affected Subcomponent: PGW
**Deployment Environment: AWS Public Cloud (orc8r), AWS outpost (AGW), AP (Mavenir)
**UE: Alcatel 4G usb dongle (https://www.alcatelmobile.com/product/mobile-broadband/mobile-wifi/linkkey-lte-cat4-usb-dongle/)

gtp_br0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST> mtu 1500
inet 192.168.128.1 netmask 255.255.255.0 broadcast 192.168.128.255
inet6 fe80::b45a:e3ff:fead:1640 prefixlen 64 scopeid 0x20
ether b6:5a:e3:ad:16:40 txqueuelen 1000 (Ethernet)
RX packets 4305385 bytes 3236553014 (3.2 GB)
RX errors 0 dropped 5 overruns 0 frame 0
TX packets 5133837 bytes 5608243745 (5.6 GB)
TX errors 0 dropped 0 overruns 0 carrier 0 collisions 0

[pcap.zip](https://github.com/magma/magma/files/7106546/pcap.zip)

**Comment by ulaskozat** [comment]
@pshelar @sudhikan what do you recommend for this?

**Reply by sudhikan** [reply]
Drop the MTU on the end device that connects to the UE to 1400bytes (or lower) I think. In general higher mtu doesn't have to be lowered for lower MTU devices... in other words, 1400 byte packet will go through un-fragmented when it hits an interface with MTU set to 1500 .. on the other hand if a 1500 byte packet were to show up at an interface with an MTU of 1400, that will get fragmented.

Also, what is the RF environment like? Since you are seeing some meaningful tput, I am not entirely sure if MTU is the issue.

**Reply by alksara** [reply]
we are observing large packets arriving from the internet at the SGi interface getting fragmented to 1416 Bytes. As our UE only supports MTU 1400, packets arriving at the UE are getting further fragmented to 1380 Bytes. To avoid further fragmentation at the UE side, we wanted to test lowering the MTU at the SGi to 1400 if possible. As this is the only difference we are seeing when comparing with the UL transmission, the UE is sending packets with 1380 Bytes size, and we are getting good throughput (~18 Mbps).

For our RF environment, we are using a Mavenir FemtoCell,RF FDD band 2, channel BW 10 MHz, 2x2 MIMO (
[S60-Datasheet-v1.1 (1).pdf](https://github.com/magma/magma/files/7121955/S60-Datasheet-v1.1.1.pdf)
), and we are using Alcatel Linkkey LTE for our UE (https://www.alcatelmobile.com/product/mobile-broadband/mobile-wifi/linkkey-lte-cat4-usb-dongle/) @sudhikan

**Comment by pshelar** [comment]
@alksara is the NATed AGW or Bridged mode?

**Reply by alksara** [reply]
@pshelar NATed 

<img width="1102" alt="Screen Shot 2021-09-08 at 10 13 07 AM" src="https://user-images.githubusercontent.com/50031454/132525830-7ba93ca0-d4f5-4499-878e-26ec3ca0d76e.png">

```
ubuntu@ip-10-65-109-201:~$ ifconfig
dhcp0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet6 fe80::4f7:73ff:feec:3336  prefixlen 64  scopeid 0x20<link>
        ether 06:f7:73:ec:33:36  txqueuelen 1000  (Ethernet)
        RX packets 16  bytes 1144 (1.1 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 818  bytes 57356 (57.3 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

envoy_cntr: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.5.0.2  netmask 255.255.255.0  broadcast 10.5.0.255
        inet6 fe80::9c82:48ff:fe51:cd68  prefixlen 64  scopeid 0x20<link>
        ether 9e:82:48:51:cd:68  txqueuelen 1000  (Ethernet)
        RX packets 323918  bytes 18150715 (18.1 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 324307  bytes 18156144 (18.1 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9001
        inet 10.65.109.201  netmask 255.255.255.0  broadcast 10.65.109.255
        inet6 fe80::6:29ff:fe79:f8fc  prefixlen 64  scopeid 0x20<link>
        ether 02:06:29:79:f8:fc  txqueuelen 1000  (Ethernet)
        RX packets 11627205  bytes 9650830470 (9.6 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 11580850  bytes 8772830778 (8.7 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.65.16.20  netmask 255.255.255.0  broadcast 10.65.16.255
        inet6 fe80::6d:b5ff:fe0e:5d32  prefixlen 64  scopeid 0x20<link>
        ether 02:6d:b5:0e:5d:32  txqueuelen 1000  (Ethernet)
        RX packets 5797796  bytes 4308855971 (4.3 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 7021484  bytes 7131510266 (7.1 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

gtp_br0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1400
        inet 192.168.128.1  netmask 255.255.255.0  broadcast 192.168.128.255
        inet6 fe80::b45a:e3ff:fead:1640  prefixlen 64  scopeid 0x20<link>
        ether b6:5a:e3:ad:16:40  txqueuelen 1000  (Ethernet)
        RX packets 5292625  bytes 3996327689 (3.9 GB)
        RX errors 0  dropped 5  overruns 0  frame 0
        TX packets 6333547  bytes 6839489586 (6.8 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

gtpu_sys_2152: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 65499
        inet6 fe80::5695:735f:2417:bd36  prefixlen 64  scopeid 0x20<link>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 1000  (UNSPEC)
        RX packets 5298948  bytes 3997531239 (3.9 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 6320900  bytes 6749568374 (6.7 GB)
        TX errors 811  dropped 615 overruns 0  carrier 0  collisions 0

ipfix0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 127.0.0.10  netmask 255.255.255.0  broadcast 127.0.0.255
        inet6 fe80::54e0:e5ff:fe31:e320  prefixlen 64  scopeid 0x20<link>
        ether 56:e0:e5:31:e3:20  txqueuelen 1000  (Ethernet)
        RX packets 19  bytes 1388 (1.3 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 818  bytes 57356 (57.3 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

li_port: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 127.1.0.0  netmask 255.255.255.0  broadcast 127.1.0.255
        inet6 fe80::88e1:95ff:fe12:a028  prefixlen 64  scopeid 0x20<link>
        ether 8a:e1:95:12:a0:28  txqueuelen 1000  (Ethernet)
        RX packets 17  bytes 1236 (1.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 816  bytes 57216 (57.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 52325562  bytes 9842201432 (9.8 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 52325562  bytes 9842201432 (9.8 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

mtr0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.1.0.1  netmask 255.255.255.0  broadcast 10.1.0.255
        inet6 fe80::2457:fcff:feb8:cb7a  prefixlen 64  scopeid 0x20<link>
        ether 26:57:fc:b8:cb:7a  txqueuelen 1000  (Ethernet)
        RX packets 18  bytes 1328 (1.3 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 816  bytes 57216 (57.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

proxy_port: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet6 fe80::c03b:70ff:fe38:56f7  prefixlen 64  scopeid 0x20<link>
        ether c2:3b:70:38:56:f7  txqueuelen 1000  (Ethernet)
        RX packets 955  bytes 68290 (68.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 830  bytes 58812 (58.8 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

uplink_br0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1400
        inet6 fe80::709c:4fff:febd:3f45  prefixlen 64  scopeid 0x20<link>
        ether 72:9c:4f:bd:3f:45  txqueuelen 1000  (Ethernet)
        RX packets 15  bytes 1068 (1.0 KB)
        RX errors 0  dropped 1  overruns 0  frame 0
        TX packets 817  bytes 57286 (57.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by alksara** [comment]
@pshelar

@arunuke mentioned to me that the team has ran into a similar issue (i.e. observing low DL throughout) and after applying the fix here (https://github.com/magma/magma/pull/9326/commits/5e0333e0f8d3b3353f087f01bc8d3344eeab6627), the issue was resolved. we would like to apply the same fix on our end, however we are on the v1.6 branch, running ovs v2.14.3 where as the patches appear to be using ovs v2.15.2. We would appreciate your kind advice on how to best apply this patch on our environment. Thanks. 

```
ubuntu@ip-10-65-109-201:~$ ovs-vswitchd --version
ovs-vswitchd (Open vSwitch) 2.14.3
```

--> magma/lte/gateway/release/build-magma.sh
```
# OVS runtime dependencies
if [[ "$OS" == "debian" ]]; then
    OVS_DEPS=(
        "magma-libfluid >= 0.1.0.6"
        "libopenvswitch >= 2.8.10"
        "openvswitch-switch >= 2.8.10"
        "openvswitch-common >= 2.8.10"
        "python-openvswitch >= 2.8.10"
        "openvswitch-datapath-module-4.9.0-9-amd64 >= 2.8.10"
        )
else
    OVS_DEPS=(
       "magma-libfluid >= 0.1.0.6"
       "libopenvswitch >= 2.14.3-7"
       "openvswitch-switch >= 2.14.3-7"
       "openvswitch-common >= 2.14.3-7"
       "openvswitch-datapath-dkms >= 2.14.3-7"
        )
```

**Reply by pshelar** [reply]
did you try running `ovs-kmod-upgrade.sh` to update OVS to '2.15.2-5'?

**Comment by alksara** [comment]
@pshelar we ran ovs-kmod-upgrade.sh and that updated OVS to 2.15.2-5 and seem to have resolved the issue. we are now observing good DL throughput ~ 70 Mbps. Thanks!

**Comment by pshelar** [comment]
As confirmed on slack running `ovs-kmod-upgrade.sh` fixes the issue.

**Comment by jpad-freedomfi** [comment]
@pshelar and @sudhikan we are testing this now for the DL issue we are seeing as well. Initial results show this fixed our issue also. Our kmods changed from 2.14.3 to 2.15.2. DL throughput jumped from 5-8Mbps to 90+Mbps. 

We should add a single UE peak throughput test to CI/CD testing.

**Comment by Bharat0908** [comment]
HI Team, 

I am running magma release over tag v1.6.1 with standalone AGW and seeing the same result of less DL throughput and good UP throughput.
UL:  300 MBPS 
DL: 20 MBPS

ifconfig of gtp_br0

vagrant@magma-dev-focal:~/magma$ ifconfig gtp_br0
gtp_br0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1464
        inet 192.168.128.1  netmask 255.255.255.0  broadcast 192.168.128.255
        inet6 fe80::b803:d9ff:fee3:4d  prefixlen 64  scopeid 0x20<link>
        ether ba:03:d9:e3:00:4d  txqueuelen 1000  (Ethernet)
        RX packets 6601  bytes 346823 (346.8 KB)
        RX errors 0  dropped 7  overruns 0  frame 0
        TX packets 27118  bytes 27502065 (27.5 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

Let me know if any other information is required.

**Reply by sudhikan** [reply]
hey @Bharat0908 any reason you are doing perf tests on a local dev setup?

**Reply by Bharat0908** [reply]
Hey @sudhikan we are just calculating the throughput which we are getting in our test lab environment.

**Reply by jpad-freedomfi** [reply]
@Bharat0908 have you used the kmod upgrade script above to make sure you have updated kmods for OVS? That is what fixed our DL throughput issue referenced above per my comment on Nov 9th.

**Reply by Bharat0908** [reply]
Hi Team,

I did the upgrade of kmod script, but still getting the same results. Execution snapshot attached below.

vagrant@magma-dev-focal:~$ sudo ./magma/lte/gateway/deploy/roles/magma/files/ovs-kmod-upgrade.sh 
This upgrade would result in datapath level downtime for few minutes!
Do you want to proceed with upgrade ?(y/n)y
Hit:1 http://archive.ubuntu.com/ubuntu focal InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal-updates InRelease                                                                            
Get:3 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]                                                                 
Hit:4 https://facebookconnectivity.jfrog.io/artifactory/list/dev-focal focal InRelease            
Get:5 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]    
Fetched 222 kB in 2s (139 kB/s)                                     
Reading package lists... Done
Building dependency tree       
Reading state information... Done
All packages are up to date.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libopenvswitch is already the newest version (2.14.3-1).
openvswitch-common is already the newest version (2.14.3-1).
openvswitch-datapath-dkms is already the newest version (2.14.3-1).
openvswitch-switch is already the newest version (2.14.3-1).
python3-openvswitch is already the newest version (2.14.3-1).
The following packages were automatically installed and are no longer required:
  libllvm11 libnetplan0 python3-colorama
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
ifdown: interface patch-up not configured
 * Detected internal interfaces: 
 * Exiting ovsdb-server (556819)
 * Starting ovsdb-server
 * Configuring Open vSwitch system IDs
 * Exiting ovs-vswitchd (556872)
 * Saving interface configuration
 * Removing vport_gtp module
 * Removing openvswitch module
 * Inserting openvswitch module
 * Starting ovs-vswitchd
 * Enabling remote OVSDB managers
 * Restoring interface configuration
 * Finding processes on dead interfaces
RTNETLINK answers: File exists
net.ipv4.ip_forward = 1
ifup: interface patch-up already configured
update successful
To cleanup control plane state run: config_stateless_agw.py flushall_redis

**Reply by Bharat0908** [reply]
I am seeing the same version of OVS upgrade. 
![image](https://user-images.githubusercontent.com/7475527/151320421-0bab38f3-9208-4edb-a5f9-ee021d877402.png)

**Comment by Bharat0908** [comment]
Hi All,

I tried updating the OVS version with kmod upgrade script at my end, but unable to do so. I am still on the version 2.14.3, which comes default with magma.
Please suggest how to upgrade it. 

Let me know if any information is needed from my end.

**Reply by pshelar** [reply]
@Bharat0908 what is version of magma agw ? I guess you are not on v1.6.1. after upgrading to 1.6.1, run the upgrade script and validate performance.

**Reply by Bharat0908** [reply]
@pshelar I am already on tag v1.6.1 and executed the script as well and still getting the low performance. Also the OVS version is still the same.
Please suggest the next steps and let me know if any other information required.

**Reply by pshelar** [reply]
As discussed on slack, the magma DEV vm does not have latest OVS. magma agw is not optimized for VM based deployments.
please use bare metal agw for perf measurements.

**Reply by Bharat0908** [reply]
Thanks @pshelar will check on bare metal server.
